Lib/fontParts/base/base.py

```diff
@@ -236,125 +307,267 @@ def raiseNotImplementedError(self):
+        the current dictionary with the contents of the `source`.
 
-    def copyData(self, source):
+        param source: The source :class:`BaseDict` instance from which
+            to copy data.
+
+        """
         super(BaseDict, self).copyData(source)
         self.update(source)
 
-    def __len__(self):
+    def __len__(self) -> int:
+        """Return the number of items in the object.
+
+        :return: An :class:`int` representing the number of dictionary items.
+
+        """
         value = self._len()
         return value
 
-    def _len(self):
-        """
-        Subclasses may override this method.
+    def _len(self) -> int:
+        """Return the number of items in the native object.
+
+        This is the environment implementation of :attr:`BaseDict.__len__`.
+
+        :return: An :class:`int` representing the number of dictionary items.
+
+        .. note::
+
+            Subclasses may override this method.
+
         """
         return len(self.keys())
 
-    def keys(self):
+    def keys(self) -> List[Any]:
+        """Return a list of keys in the object.
+
+        :return: A :class:`list` of dictionary keys.
+
+        """
         keys = self._keys()
         if self.keyNormalizer is not None:
             keys = [self.keyNormalizer.__func__(key) for key in keys]
         return keys
 
-    def _keys(self):
-        """
-        Subclasses may override this method.
+    def _keys(self) -> List[Any]:
+        """Return a list of keys in the native object.
+
+        This is the environment implementation of :meth:`BaseDict.keys`.
+
+        :return: A :class:`list` of dictionary keys. If
+            a :cvar:`BaseDict.keyNormalizer` is set, it will have been applied
+            to each key.
+
+        .. note::
+
+            Subclasses may override this method.
+
         """
         return [k for k, v in self.items()]
 
-    def items(self):
+    def items(self) -> List[Tuple[Any, Any]]:
+        """Return a list of key-value pairs in the object.
+
+        :return: A :class:`list` of :class:`tuple` items containing key-value pairs.
+
+        """
         items = self._items()
         if self.keyNormalizer is not None and self.valueNormalizer is not None:
-            values = [
+            items = [
                 (self.keyNormalizer.__func__(key),
                  self.valueNormalizer.__func__(value))
                 for (key, value) in items
             ]
-        return values
+        return items
+
+    def _items(self) -> List[Tuple[Any, Any]]:
+        """Return a list of key-value pairs in the native object.
+
+        This is the environment implementation of :meth:`BaseDict.items`.
+
+        :return: A :class:`list` of :class:`tuple` items containing key-value pairs.
+            If both :cvar:`BaseDict.keyNormalizer` and :cvar:`BaseDict.valueNormalizer`
+            are set, they will have been applied to the keys and values, respectively.
```

Comment:
better to say `they will be applied in the calling method to the key and values`

Reply:
Changed.